designs/v1alpha4-api.md

```diff
@@ -0,0 +1,130 @@
+
+1. (+) Parameters that are natural siblings (e.g. `zones`, `securityGroups`) will be no longer be separated by `provider`.
+2. (-) It's not immediately clear to readers which parameters are vendor specific.
+3. (-) Key name conflicts may arise between provider and vendor fields.
+4. (-) `kind` and `apiVersion` are awkwardly placed at the top level of spec.
+
+### Alternative: Duck Typing
+
+It's possible to follow [Knative's Duck Typing approach](https://www.youtube.com/watch?v=kldVg63Utuw) and build vendor specific CRDs that contain vendor neutral snippets (or ducks) that Karpenter can recognize. Cloud Providers would consume Karpenter generic controllers as libaries, which would behave against the generic API snippets. For example:
+
+```bash
+apiVersion: karpenter.k8s.aws/v1alpha1 # Vendor Specific
+kind: Provisioner
+spec:
+  limits: {} # Vendor neutral, recognized by Karpenter generic controllers
+  labels: {} # Vendor neutral, recognized by Karpenter generic controllers
+  taints: [] # Vendor netural, recognized by Karpenter generic controllers
+  subnets: [] # Vendor specific, only recognized by AWS Cloud Provider code
+```
+
+1. (+) Versioning is only defined in one place
+2. (+) Vendors may change versions decoupled from each other.
+3. (+) Parameters that are natural siblings (e.g. `zones`, `securityGroups`) will be no longer be separated by `provider`.
+4. (-) Cloud Provider must do more than simply implementing an interface (they must define an CR, etc).
+
+This approach is quite attractive, but requires significant effort and changes to the project.
+
+## Pluralizion for all Constraints
+We expand the constraints of a Provisioner from a scalar to a vector in all cases. This will apply to both `operatingSystem` and `architecture` enabling the operator to specify greater flexibility. This supports use cases such as heterogenous architectures within a single provisioner, selected dynamically at runtime by the cloud provider. More importantly, this change creates a consistent and predictable semantic for all vendor neutral constraints.
+### Example
+```bash
+spec:
+  operatingSystem: ["linux", "windows"]
+  architecture: ["amd64", "arm64"]
+```
+
+## Limits
+
+We introduce a new field `spec.limits` that contains configuration parameters to limit scaling to control costs.
+
+### Example
+```bash
+spec:
+  limits:
+    ready: 500 # Flat or Percentage. Karpenter will not launch additional capacity if ready nodes exceeds this value
+    unready: 20% # Flat or Percentage. Karpenter will not launch additional capacity if current unready nodes exceeds this value
+    resources: # Karpenter will not launch additional capacity if current capacity exceeds this value
+      cpu: 1000
+      memory: 1000Gi
+```
+
+## Reparenting top level fields
+
+With the introduction of additional top level fields, top level parameters no longer carry the same conceptual weight. To address this, we will add parent fields for `constraints` and `scale down`.
```

Comment:
I'm having trouble parsing this... why won't they carry the same "conceptual weight?"